storage/blockchain/src/service/read.rs

```diff
@@ -32,156 +29,34 @@ use crate::{
-            .build()
-            .unwrap();
-
-        // Create a semaphore with the same amount of
-        // permits as the amount of reader threads.
-        let semaphore = PollSemaphore::new(Arc::new(Semaphore::new(reader_count)));
-
-        // Return a handle to the pool.
-        Self {
-            pool: Arc::new(pool),
-            semaphore,
-            permit: None,
-            env: Arc::clone(env),
-        }
-    }
-
-    /// Access to the actual database environment.
-    ///
-    /// # ⚠️ Warning
-    /// This function gives you access to the actual
-    /// underlying database connected to by `self`.
-    ///
-    /// I.e. it allows you to read/write data _directly_
-    /// instead of going through a request.
-    ///
-    /// Be warned that using the database directly
-    /// in this manner has not been tested.
-    #[inline]
-    pub const fn env(&self) -> &Arc<ConcreteEnv> {
-        &self.env
-    }
-}
-
-impl tower::Service<BCReadRequest> for DatabaseReadHandle {
-    type Response = BCResponse;
-    type Error = RuntimeError;
-    type Future = ResponseReceiver;
-
-    #[inline]
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        // Check if we already have a permit.
-        if self.permit.is_some() {
-            return Poll::Ready(Ok(()));
-        }
-
-        // Acquire a permit before returning `Ready`.
-        let permit =
-            ready!(self.semaphore.poll_acquire(cx)).expect("this semaphore is never closed");
-
-        self.permit = Some(permit);
-        Poll::Ready(Ok(()))
-    }
-
-    #[inline]
-    fn call(&mut self, request: BCReadRequest) -> Self::Future {
-        let permit = self
-            .permit
-            .take()
-            .expect("poll_ready() should have acquire a permit before calling call()");
-
-        // Response channel we `.await` on.
-        let (response_sender, receiver) = oneshot::channel();
-
-        // Spawn the request in the rayon DB thread-pool.
-        //
-        // Note that this uses `self.pool` instead of `rayon::spawn`
-        // such that any `rayon` parallel code that runs within
-        // the passed closure uses the same `rayon` threadpool.
-        //
-        // INVARIANT:
-        // The below `DatabaseReader` function impl block relies on this behavior.
-        let env = Arc::clone(&self.env);
-        self.pool.spawn(move || {
-            let _permit: OwnedSemaphorePermit = permit;
-            map_request(&env, request, response_sender);
-        }); // drop(permit/env);
-
-        InfallibleOneshotReceiver::from(receiver)
-    }
+/// Initialize the blockchain database read service, with a specific rayon thread-pool instead of
+/// creating a new one.
+///
+/// Should be called _once_ per actual database.
+#[cold]
+#[inline(never)] // Only called once.
+pub fn init_read_service_with_pool(env: Arc<ConcreteEnv>, pool: Arc<ThreadPool>) -> BCReadHandle {
```

Comment:
I added the ability for users to specify a thread pool here instead of always creating a new one, this will allow sharing the reader thread pool between the blockchain & txpool